### ISSUE TITLE
'make test' reports failure; everything is "Not run"

### ISSUE THREAD

**Issue by ExpHP** [dftbplus/dftbplus#244]
Attempted on both github master (be6fde2c717a5577) and the 18.2 release.

---

On my machine, `make test` emits a whole bunch of lines with the word `TODO`, followed by a summary which appears to suggest that no tests were run, and a nonzero exit code.  Is the testing code implemented?  I'm not sure what to make of this...

Similar results occur for `test_api` on master.

---

```
$ utils/get_opt_externals
$ make
$ make test
mkdir -p /home/lampam/Downloads/dftbplus-18.2/_build
[ -r /home/lampam/Downloads/dftbplus-18.2/RELEASE ] && cp -a /home/lampam/Downloads/dftbplus-18.2/RELEASE /home/lampam/Downloads/dftbplus-18.2/_build/RELEASE \
        || /home/lampam/Downloads/dftbplus-18.2/utils/build/update_release /home/lampam/Downloads/dftbplus-18.2/_build/RELEASE \
        || echo "(UNKNOWN RELEASE)" > /home/lampam/Downloads/dftbplus-18.2/_build/RELEASE
mkdir -p /home/lampam/Downloads/dftbplus-18.2/_build/external/xmlf90
make -C /home/lampam/Downloads/dftbplus-18.2/_build/external/xmlf90 \
          -f /home/lampam/Downloads/dftbplus-18.2/external/xmlf90/make.dpbuild \
          ROOT=/home/lampam/Downloads/dftbplus-18.2 BUILDROOT=/home/lampam/Downloads/dftbplus-18.2/_build
make[1]: Entering directory '/home/lampam/Downloads/dftbplus-18.2/_build/external/xmlf90'
touch -r libxmlf90.a /home/lampam/Downloads/dftbplus-18.2/_build/external/xmlf90/BUILD_TIMESTAMP
make[1]: Leaving directory '/home/lampam/Downloads/dftbplus-18.2/_build/external/xmlf90'
mkdir -p /home/lampam/Downloads/dftbplus-18.2/_build/external/fsockets
make -C /home/lampam/Downloads/dftbplus-18.2/_build/external/fsockets \
          -f /home/lampam/Downloads/dftbplus-18.2/external/fsockets/make.dpbuild \
          ROOT=/home/lampam/Downloads/dftbplus-18.2 BUILDROOT=/home/lampam/Downloads/dftbplus-18.2/_build
make[1]: Entering directory '/home/lampam/Downloads/dftbplus-18.2/_build/external/fsockets'
touch -r libfsockets.a /home/lampam/Downloads/dftbplus-18.2/_build/external/fsockets/BUILD_TIMESTAMP
make[1]: Leaving directory '/home/lampam/Downloads/dftbplus-18.2/_build/external/fsockets'
mkdir -p /home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+
make -C /home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+ -f /home/lampam/Downloads/dftbplus-18.2/prog/dftb+/make.build \
    ROOT=/home/lampam/Downloads/dftbplus-18.2 BUILDROOT=/home/lampam/Downloads/dftbplus-18.2/_build
make[1]: Entering directory '/home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+'
touch -r /home/lampam/Downloads/dftbplus-18.2/_build/external/xmlf90/BUILD_TIMESTAMP _extlib_xmlf90
touch -r /home/lampam/Downloads/dftbplus-18.2/_build/external/fsockets/BUILD_TIMESTAMP _extlib_fsockets
make[1]: Leaving directory '/home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+'
make -C /home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+ \
    -f /home/lampam/Downloads/dftbplus-18.2/prog/dftb+/make.build \
    ROOT=/home/lampam/Downloads/dftbplus-18.2 BUILDROOT=/home/lampam/Downloads/dftbplus-18.2/_build test
make[1]: Entering directory '/home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+'
derivatives/Si_2_Delta:  preparing .. running .. comparing .. TODO.
legacy/Si2_oldSKinterp:  preparing .. running .. comparing .. TODO.
legacy/Si2_polyRep:      preparing .. running .. comparing .. TODO.
non-scc/Si_2:    preparing .. running .. comparing .. TODO.
non-scc/Si_2_independentk:       preparing .. running .. comparing .. TODO.
scc/H2O2_3rdfull-damp:   preparing .. running .. comparing .. TODO.
scc/H3:  preparing .. running .. comparing .. TODO.
spin/H2:         preparing .. running .. comparing .. TODO.
analysis/C2H4_localise:  preparing .. running .. comparing .. TODO.
analysis/graphene_localise:      preparing .. running .. comparing .. TODO.
dftb+u/CH3:      preparing .. running .. comparing .. TODO.
dispersion/2H2O:         preparing .. running .. comparing .. TODO.
dispersion/2H2O_uff:     preparing .. running .. comparing .. TODO.
geoopt/H2O-nonscc:       preparing .. running .. comparing .. TODO.
non-scc/CH4:     preparing .. running .. comparing .. TODO.
non-scc/decapentaene:    preparing .. running .. comparing .. TODO.
scc/2H2O-3rdorder:       preparing .. running .. comparing .. TODO.
scc/2H2O-3rdorder_read:  preparing .. running .. comparing .. TODO.
scc/C2H6_3rdfull:        preparing .. running .. comparing .. TODO.
scc/C2H6_3rdfull-damp:   preparing .. running .. comparing .. TODO.
scc/H2O2_3rdfull:        preparing .. running .. comparing .. TODO.
scc/H2O2-3rdfull-ldep:   preparing .. running .. comparing .. TODO.
scc/H2O-extchrg:         preparing .. running .. comparing .. TODO.
scc/H2O-extchrg-blur:    preparing .. running .. comparing .. TODO.
scc/H2O-extchrg-direct:  preparing .. running .. comparing .. TODO.
scc/H2O-extchrg-periodic:        preparing .. running .. comparing .. TODO.
scc/H2O-extfield:        preparing .. running .. comparing .. TODO.
spin/H2O-periodic:       preparing .. running .. comparing .. TODO.
spin/OH_commonFermi:     preparing .. running .. comparing .. TODO.
derivatives/Si_2_Richardson:     preparing .. running .. comparing .. TODO.

                            (snip)

md/SiC64-xlbomdfast:     preparing .. running .. comparing .. TODO.
timedep/C66O10N4H44_OscWindow:   preparing .. running .. comparing .. TODO.
timedep/C60_OscWindow:   preparing .. running .. comparing .. TODO.
timedep/C60_EandOsc:     preparing .. running .. comparing .. TODO.
non-scc/Si_216:  preparing .. running .. comparing .. TODO.
md/ptcda-xlbomd-ldep:    preparing .. running .. comparing .. TODO.
geoopt/Vsi+O_lbfgs:      preparing .. running .. comparing .. TODO.
geoopt/Vsi+O:    preparing .. running .. comparing .. TODO.
==============================================================================
TEST SUMMARY
------------------------------------------------------------------------------
Match:
    None

Not run:
    derivatives/Si_2_Delta              analysis/Fe2_antiferromagnetic
    legacy/Si2_oldSKinterp              non-scc/10-0Ctube_Efield
    legacy/Si2_polyRep                  non-scc/10-10Ctube
    non-scc/Si_2                        analysis/10-0Ctube_ESP
    non-scc/Si_2_independentk           md/H2O-extfield
    scc/H2O2_3rdfull-damp               non-scc/10-0Ctube
    scc/H3                              spin/Fe4_noncolinear
    spin/H2                             non-scc/HBDI-neutral
    analysis/C2H4_localise              non-scc/HBDI-cationic
    analysis/graphene_localise          scc/C60
    dftb+u/CH3                          analysis/C60_ESP
    dispersion/2H2O                     spin/H2O
    dispersion/2H2O_uff                 timedep/cyclopentadienyl
    geoopt/H2O-nonscc                   timedep/C6H6-Sym
    non-scc/CH4                         timedep/C6H6-Sym_Arnoldi
    non-scc/decapentaene                timedep/OCH2-S1-Opt
    scc/2H2O-3rdorder                   h-bonds/H5_defaults
    scc/2H2O-3rdorder_read              h-bonds/H5_ONS_forces
    scc/C2H6_3rdfull                    h-bonds/H5_ONS_forces_eq
    scc/C2H6_3rdfull-damp               md/Si_8_restart
    scc/H2O2_3rdfull                    spin/GaAs_2
    scc/H2O2-3rdfull-ldep               spin/GaAs-spin-ext
    scc/H2O-extchrg                     spinorbit/As4S4
    scc/H2O-extchrg-blur                dispersion/DNA-damped
    scc/H2O-extchrg-direct              md/ice_Ic
    scc/H2O-extchrg-periodic            derivatives/C6H6_scc
    scc/H2O-extfield                    dftb+u/GaAs_2
    spin/H2O-periodic                   geoopt/diamond_presure
    spin/OH_commonFermi                 scc/C60_Fermi
    derivatives/Si_2_Richardson         scc/10-0Ctube-extfield
    geoopt/H2O-constr                   scc/SiC_64
    legacy/SiC_polyRep                  scc/SiC_32-extchrg-blur
    md/SiH-surface_restart              md/H3
    scc/C4H8_3rdfull                    sockets/diamond
    scc/C4H8_3rdfull-damp               sockets/diamond_exit
    scc/CH2_n_3rdfull                   sockets/H2O
    scc/CH2_n_3rdfull_damp              sockets/H2O_cluster
    scc/GaAs_2_restart                  spinorbit/GaAs_2
    spin/Fe4_commonFermi                geoopt/Si_2_latconst
    timedep/2CH3-Temp                   md/SiH-surface
    timedep/2CH3-Triplet-Temp           geoopt/Si_2_lattice_lbfgs
    timedep/NO                          geoopt/Si_2_lattice
    timedep/propadiene_OscWindow        md/Si_8_NHC
    analysis/H2O_ESP                    timedep/C66O10N4H44_Ewindow
    analysis/H2O_mdESP                  spin/Fe4_Fermi
    geoopt/H2O_lbfgs                    non-scc/Si41C23N35
    geoopt/H2O                          geoopt/diamond_isotropic
    md/Si_8_NHC_restart                 spinorbit/EuN
    non-scc/GaAs_2                      geoopt/Si_6
    scc/GaAs_2                          md/ptcda-xlbomdfast
    scc/GaAs_2_customU                  spinorbit/EuN_customU
    spinorbit/Fe2_dual                  dispersion/DNA
    spinorbit/Fe2_dual_field            spinorbit/Si2_dual
    spinorbit/Si_2                      md/Si_8
    dispersion/DNA_uff                  md/Si_8-thermostat2
    timedep/C4H6-S1-Force               geoopt/GaAs_8_latconst_lbfgs
    timedep/C4H6-Singlet                geoopt/GaAs_8_latconst
    timedep/C4H6-Singlet_wfn            analysis/Ga4As4_ESP
    timedep/C4H6-T1-Force               md/Si_8-thermostat
    timedep/C4H6-Triplet                md/ptcda-xlbomdfast-ldep
    dftb+u/Fe4                          md/Si_8-tempprofile
    dftb+u/Fe4_read                     geoopt/Vsi+O-nonscc
    geoopt/Cchain_lattice_lbfgs         non-scc/Si_384
    geoopt/Cchain_lattice               md/DNA
    scc/H2O+CH3COOH-3rdorder            md/DNA_Berendsen2
    spinorbit/Fe2                       md/ptcda-xlbomd
    md/SiC64-xlbomdfast-T0              non-scc/Si_216
    md/SiC64-xlbomdfast                 md/ptcda-xlbomd-ldep
    timedep/C66O10N4H44_OscWindow       geoopt/Vsi+O_lbfgs
    timedep/C60_OscWindow               geoopt/Vsi+O
    timedep/C60_EandOsc
------------------------------------------------------------------------------
Status: FAIL
------------------------------------------------------------------------------
Details in:
    _autotest/stderror.log
    _autotest/tagdiff.log
==============================================================================
make[1]: *** [/home/lampam/Downloads/dftbplus-18.2/prog/dftb+/make.build:76: test] Error 1
make[1]: Leaving directory '/home/lampam/Downloads/dftbplus-18.2/_build/prog/dftb+'
make: *** [makefile:92: test_dftb+] Error 2
```

The log files mentioned `_build/prog/dftb+/_autotest/{stderror,tagdiff}.log` are both empty.

**Comment by bhourahine** [comment]
Assuming the compilation was successful, the required files to run tests might be missing. Did you run the script utils/get_opt_externals  ?

**Comment by ExpHP** [comment]
Yes.  When they are not present it produces different errors which are clearly about missing ".skf" files, rather than the vague errors shown here.

**Comment by bhourahine** [comment]
OK, so what is in _build/prog/dftb+/_autotest/stderror.log ?

**Comment by ExpHP** [comment]
Empty, as mentioned in the OP.

**Comment by bhourahine** [comment]
OK, so what is in (for example) 
_build/prog/dftb+/_autotest/non-scc/CH4/output
and
_build/prog/dftb+/_autotest/non-scc/CH4/stderror.log
?

**Comment by ExpHP** [comment]
Ooh!  Now we're getting somewhere:

**`_build/prog/dftb+/_autotest/non-scc/CH4/output`**: empty

**`_build/prog/dftb+/_autotest/non-scc/CH4/stderror.log`**:

```
/home/lampam/Downloads/dftbplus-18.2/test/prog/dftb+/bin/autotest2: line 338: time: command not found
```

And it's true!  Apparently the base Arch Linux installation does not come with `/usr/bin/time`.  I never even noticed it was missing since bash has its own `time` keyword.

After installing the `time` package, the tests run properly.

**Comment by ExpHP** [comment]
Since things are working for me now, if you don't feel there's anything actionable here, you can close this.

To me it looks like the possible actionable items here are:

* The tester could be made robust to not having `time`. (or explicitly error without it)
* Maybe "failure to run a test" could somehow be detected as a failure rather than "Not run."
* The tester could mention the existence of those log files that exist for individual testcases, allowing users to debug this issue on their own.

IMO, just the last bullet would be enough.  Even knowing that it is missing from Arch, expecting `time` to exist doesn't seem unreasonable to me.

**Comment by bhourahine** [comment]
The time command is now addressed in #255

**Comment by ExpHP** [comment]
Fix confirmed on my end.  Thanks!
